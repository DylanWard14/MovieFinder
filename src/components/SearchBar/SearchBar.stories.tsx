import React from 'react';
import styled from 'styled-components/macro';

import { SearchStoreState } from '../../store/Search/types';
import SearchBar from './SearchBar';

const Wrapper = styled.div`
    width: 100%;
    height: 100%;
    background: black;
`;

export default {
    title: 'Components/SearchBar',
    decorators: [(storyFn: any) => <Wrapper>{storyFn()}</Wrapper>],
};

const searchStore: SearchStoreState = {
    searchTerm: '',
    data: undefined,
    loading: false,
};

const onSubmit = (searchTerm: string) => console.log(`searching for ${searchTerm}`);

export const DefaultExample = () => <SearchBar searchStore={searchStore} handleSearchSubmit={onSubmit} />;

export const WithResultsExample = () => (
    <SearchBar
        searchStore={{
            ...searchStore,
            searchTerm: 'result',
            data: [{ title: 'result1' }, { title: 'result2' }, { title: 'result3' }],
        }}
        handleSearchSubmit={onSubmit}
    />
);

export const WithNoDataExample = () => (
    <SearchBar searchStore={{ ...searchStore, searchTerm: 'result', data: [] }} handleSearchSubmit={onSubmit} />
);
