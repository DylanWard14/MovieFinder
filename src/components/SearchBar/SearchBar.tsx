import React, { useState } from 'react';
import styled from 'styled-components/macro';
import { Manager, Reference, Popper } from 'react-popper';
import { Link } from 'react-router-dom';

import { SearchStoreState } from '../../store/Search/Types';
import useModalToggle from '../../hooks/useModalToggle';

const StyledForm = styled.form`
    width: 100%;
`;

const StyledInput = styled.input`
    width: 100%;
    background: none;
    color: white;
    border: none;
    border-bottom: 1px solid white;
    font-size: 1rem;
    height: 1.5625rem;
    box-sizing: border-box;
    padding: 0.5rem 0.5rem;
`;

const StyledPopper = styled.div`
    width: 100%;
    background: black;
    z-index: 100;
`;

const Results = styled.ul`
    width: 100%;
    background: black;
    z-index: 100;
    list-style-type: none;
    padding: 0;
    margin: 0.5rem 0;
    position: absolute;
    color: ${props => props.theme.colors.secondary};
`;

const Result = styled(Link)`
    width: 100%;
    display: inline-block;
    box-sizing: border-box;
    padding: 0.5rem 1rem;
    text-decoration: none;
    color: inherit;
    font-size: 1.125rem;

    :hover,
    :focus {
        background: ${props => props.theme.colors.main};
        color: black;
        cursor: pointer;
    }
`;

const NoData = styled.li`
    padding: 0.5rem 1rem;
    font-size: 1.125rem;
`;

interface SearchBarProps {
    searchStore: SearchStoreState;
    handleSearchSubmit: (searchTerm: string) => void;
    className?: string;
}

const SearchBar: React.FunctionComponent<SearchBarProps> = ({ searchStore, handleSearchSubmit, className }) => {
    const [searchTerm, setSearchTerm] = useState('');
    const handleSubmit = (event: React.FormEvent<EventTarget>) => {
        event.preventDefault();
        handleSearchSubmit(searchTerm);
    };

    const { ref: ModalRef, visible } = useModalToggle(true);

    return (
        <StyledForm onSubmit={handleSubmit} ref={ModalRef} className={className}>
            <Manager>
                <Reference>
                    {({ ref }) => (
                        <StyledInput
                            type="text"
                            id="search"
                            name="search"
                            placeholder="Search For Movie Here..."
                            autoComplete="off"
                            ref={ref}
                            value={searchTerm}
                            onChange={(event: React.FormEvent<HTMLInputElement>) => {
                                setSearchTerm(event.currentTarget.value);
                            }}
                        />
                    )}
                </Reference>
                {visible && !searchStore.loading && (
                    <Popper placement="bottom">
                        {({ ref, style, placement, arrowProps }) => (
                            <StyledPopper ref={ref} style={style} data-placement={placement} tabIndex={1}>
                                <Results>
                                    {searchStore?.data?.length
                                        ? searchStore.data.map((item: any) => {
                                              return (
                                                  <li tabIndex={0} key={`${item.title}_Search_Result`}>
                                                      <Result to={`/${item.id}`}>{item.title}</Result>
                                                  </li>
                                              );
                                          })
                                        : searchStore.searchTerm && <NoData>No results found...</NoData>}
                                </Results>
                                <div ref={arrowProps.ref} style={arrowProps.style} />
                            </StyledPopper>
                        )}
                    </Popper>
                )}
            </Manager>
        </StyledForm>
    );
};

export default SearchBar;
