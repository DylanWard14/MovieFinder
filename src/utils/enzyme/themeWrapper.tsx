import React from 'react';
import { render, shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import myTheme from '../../theme';

export const renderWrapper = (component: React.ReactNode) => {
    return render(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};

export const shallowWrapper = (component: React.ReactNode) => {
    return shallow(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};

export const mountWrapper = (component: React.ReactNode) => {
    return mount(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};
