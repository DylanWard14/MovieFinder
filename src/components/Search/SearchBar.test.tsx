import React from 'react';

import { mountWrapper, renderWrapper } from '../../utils/enzyme/themeWrapper';
import SearchBar from './SearchBar';

jest.mock('popper.js', () => {
    const PopperJS = jest.requireActual('popper.js');
    class Popper {
        static placements = PopperJS.placements;
        constructor() {
            return {
                destroy: () => {},
                scheduleUpdate: () => {},
                placements: null,
            };
        }
    }
    return Popper;
});

describe('SearchBar', () => {
    const searchStore = {
        loading: false,
        searchTerm: '',
        data: undefined,
        error: undefined,
    };
    const searchSubmit = jest.fn();
    it('renders correctly', () => {
        const wrapper = renderWrapper(<SearchBar searchStore={searchStore} handleSearchSubmit={searchSubmit} />);
        expect(wrapper).toMatchSnapshot();
    });
    it('handleSearchSubmit is not called if there is no search term', () => {
        const wrapper = mountWrapper(<SearchBar searchStore={searchStore} handleSearchSubmit={searchSubmit} />);
        wrapper.find('search-form').simulate('submit');
        expect(searchSubmit).toHaveBeenCalledTimes(0);
    });

    it('handleSearchSubmit is called with the searchTerm', () => {
        const wrapper = mountWrapper(
            <SearchBar searchStore={{ ...searchStore, searchTerm: 'testTerm' }} handleSearchSubmit={searchSubmit} />,
        );
        wrapper.find('search-form').simulate('submit');
        expect(searchSubmit).toHaveBeenCalledTimes(1);
        expect(searchSubmit).toHaveBeenCalledWith('testTerm');
    });
});
