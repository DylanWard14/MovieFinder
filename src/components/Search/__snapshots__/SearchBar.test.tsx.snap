// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders correctly 1`] = `
<form
  class="SearchBar__StyledForm-sc-196gn3w-0 iBTxGs"
  id="search_form"
>
  <input
    autocomplete="off"
    class="SearchBar__StyledInput-sc-196gn3w-1 kLlndz"
    id="search"
    name="search"
    placeholder="Search For Movie Here..."
    type="text"
    value=""
  />
  <div
    class="SearchBar__StyledPopper-sc-196gn3w-2 hJvTep"
    style="position:absolute;top:0;left:0;opacity:0;pointer-events:none"
    tabindex="1"
  >
    <ul
      class="SearchResults__Results-miypjy-0 iVYBlS"
      id="search_results"
    />
    <div />
  </div>
</form>
`;
